### PR TITLE
Issue 723

### PR DESCRIPTION
**Pull request by chrisdennis**
Wiring the clustered cache API to the voltron client code.

### CODE REVIEW

**Comment by ljacomet** [comment]
Pretty sure that the empty file `clustered/client/private/cache/retriever/catalog.xml` should not be part of this

File: clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java [hunk 1]
```diff
@@ -34,46 +48,113 @@
...
+    entityFactory = new EhcacheClientEntityFactory(clusterConnection);
+    if (!entityFactory.acquireLeadership(entityIdentifier)) {
+      throw new IllegalStateException("Couldn't acquire cluster-wide maintenance lease");
+    }
   }
 
   @Override
   public void stop() {
-    // TODO: Implement stop
+    entityFactory.abandonLeadership(entityIdentifier);
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java]: Is that safe to call when not holding it?

File: clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java [hunk 2]
```diff
@@ -34,46 +48,113 @@
...
+    try {
+      clusterConnection.close();
+    } catch (IOException ex) {
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java]: No need to clear the entityFactory?

File: core/src/main/java/org/ehcache/core/EhcacheManager.java [hunk 3]
```diff
@@ -117,6 +117,7 @@ public EhcacheManager(Configuration config, Collection<Service> services, boolea
     this.serviceLocator = new ServiceLocator(services.toArray(new Service[services.size()]));
     this.useLoaderInAtomics = useLoaderInAtomics;
     validateServicesConfigs();
+    resolveServices();
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/EhcacheManager.java]: Why is this needed at this time?

Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/EhcacheManager.java]: So that the full set of services is loaded in to the ServiceLocator before maintenance is called.  Before this change it wasn't being done until init, which means maintenance won't see the whole set.

File: clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java (hunk 1 above)
Comment (chrisdennis) [inline clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java]: Yes... I believe so but I will verify this again.

- [ ] Verify it's safe to abandon leadership when not owning it (with a test maybe?)

File: clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java (hunk 2 above)
Comment (chrisdennis) [inline clustered/client/src/main/java/org/ehcache/clustered/service/DefaultClusteringService.java]: That would probably be a good move, will make the change.

File: core/src/main/java/org/ehcache/core/EhcacheManager.java [hunk 4]
```diff
@@ -523,24 +537,8 @@ public void deregisterListener(CacheManagerListener listener) {
...
-          throw new IllegalArgumentException("Couldn't resolve Service " + serviceConfig.getServiceType().getName());
-        }
-      }
-      serviceLocator.loadDependenciesOf(ServiceDeps.class);
-      try {
-        serviceLocator.startAllServices();
-      } catch (Exception e) {
-        throw st.failed(e);
-      }
+      serviceLocator.startAllServices();
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/EhcacheManager.java]: Removing the surrounding catch block will cause an exception to be triggered when trying to stop in the lower catch block which will end up shadowing the original exception.
So AFAICT the try/catch block is required here.

File: core/src/main/java/org/ehcache/core/StatusTransitioner.java [hunk 5]
```diff
@@ -30,7 +30,7 @@
 /**
  * @author Alex Snaps
  */
-final class StatusTransitioner {
+public final class StatusTransitioner {
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/StatusTransitioner.java]: None of these visibility changes are required. Please revert.

File: core/src/main/java/org/ehcache/core/EhcacheManager.java (hunk 3 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/EhcacheManager.java]: Indeed - makes sense

File: core/src/main/java/org/ehcache/core/StatusTransitioner.java (hunk 5 above)
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/StatusTransitioner.java]: Yeah... thats gunk pulled across from earlier hackery... will revert.

File: core/src/main/java/org/ehcache/core/EhcacheManager.java (hunk 4 above)
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/EhcacheManager.java]: Confused... let's chat about this.

Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/EhcacheManager.java]: Sorry, reviewer stupid
